packages/web-components/fast-html/RENDERING.md

```diff
@@ -0,0 +1,311 @@
+    text: string;
+}
+
+export interface IInitialStateService {
+  init(): InitialState;
+}
+
+export class InitialStateService implements IInitialStateService {
+    private initialState: any;
+
+    constructor() {
+        this.initialState = window.__INITIAL_STATE__;
+    }
+
+    public init(...keys: Array<string>): InitialState {
+        const config: any = {};
+
+        for (let key of keys) {
+            config[key] = this.initialState[key];
+        }
+
+        return config;
+    }
+}
+
+export const initialStateFactory = DI.createContext<IInitialStateService>(
+  'i-initial-state-service',
+  (x) => x.singleton(InitialStateService)
+);
+```
+
+Include the DI service to your component.
+
+Example:
+```typescript
+import { attr, FASTElement, Observable } from "@microsoft/fast-element";
+import {
+    initialStateFactory,
+    InitialStateService,
+    type IInitialStateService,
+} from "./initial-state.js";
+import { inject } from '@microsoft/fast-element/di.js';
+
+export class MyComponent extends FASTElement {
+    @inject(initialStateFactory) initialStateService!: InitialStateService;
+
+    @attr
+    text: string = "";
+
+    @attr({
+        attribute: "defer-hydration",
+        mode: "boolean"
+    })
+    deferHydration?: boolean;
+
+    service!: IInitialStateService;
+
+    private shadowOptions = false;
+    public initialState = false;
+
+    connectedCallback() {
+        super.connectedCallback();
+
+        Observable.defineProperty(this.$fastController.definition, "shadowOptions");
+
+        Observable.getNotifier(this.$fastController.definition).subscribe(
+            {
+                handleChange: (value) => {
+                    if (value.shadowOptions) {
+                        this.shadowOptions = true;
+                        this.hasStateAndShadowOptions();
+                    }
+                },
+            },
+            "shadowOptions"
+        );
+
+        this.loadItemFromInitialState();
+    }
+
+    hasStateAndShadowOptions() {
+        if (this.shadowOptions && this.initialState) {
+            this.deferHydration = false;
+        }
+    }
+
+    loadItemFromInitialState(): void {
+        const initialState = this.initialStateService.init(["text"]);
+
+        this.text = initialState.text;
+
+        this.initialState = true;
+
+        this.hasStateAndShadowOptions();
+    }
+}
+
+MyComponent.define({
+    name: "my-component",
+    shadowOptions: null
+});
+```
+
+Notice that there is a `defer-hydration` attribute that gets removed once initial state has been applied and `shadowOptions` has also been defined. This allows us to ensure that the template has been found and parsed to the component and the initial state has been applied before we connect the component to the rendered HTML.
+
+Depending on your application structure it might be more maintainable to create an abstract class that extends `FASTElement` to re-use the service logic and the removal of `defer-hydration`.
```

Comment:
Could also leave a note here mentioning that `define` can be overridden to provide `null` as a default value for `shadowOptions` which may be desirable especially for folks that have been using `fast-element` for a long time.

Reply:
This note could also be added to the README where setting to null is mentioned to prevent FUOC